New Introduction Docs Page (#1538)
zenml-io/zenml: docs/book/introduction.md

Reviewer: I feel like these questions here need a change, or need to refined towards our persona: 

🤹 Are you an ML engineer or data scientist shipping models to production while juggling a plethora of tools?
🤷‍♂️ Do you struggle with versioning data, code, and models in your projects?
👀 Have you had trouble replicating production pipelines and monitoring models in production?
[🙇](https://apps.timwhitlock.info/emoji/tables/unicode#emoji-modal) Are you dealing with lots of unstructured notebooks or scripts with lots of manual processes?
[🚧](https://apps.timwhitlock.info/emoji/tables/unicode#emoji-modal) Do you find it challenging to transition your ML workflows from the experimentation setting to a production-ready environment?

Maybe only 3 challenges and more focus on our top 3 things we solve?

Author: IMO this seems like boilerplate text, I took the problem list out entirely, let's just get to the point

# ⭐ Introduction

ZenML is an extensible, open-source MLOps framework for creating portable, 
production-ready machine learning pipelines.
By decoupling infrastructure from code, ZenML gives all developers in your 

> organization more freedom and independence in how they approach their work.

Reviewer 2: "gives freedom and independence" feels a little vague. This is an important second sentence where we give the 'why' / value prop for ZenML so it should be super tight. (Sorry no immediate suggestions come to mind on how to rephrase, but I think it needs work).

Author: Okay let's see if you like the new version better

![ZenML Overview](.gitbook/assets/intro_zenml_overview.png)

{% tabs %}

{% tab title="ZenML for Data Scientists" %}

> As a Data Scientist, ZenML gives you the freedom to focus on modeling and

Reviewer 2: Maybe it's just me, but whenever I see 'As a ...' sentences, I think of 'As a large language model trained by OpenAI...'. Smells a bit odd, but maybe it's just me...

Author: @bcdurak didn't like it either, see above. I'll change it 👍

experimentation while writing code that is immediately production-ready.

**Develop Locally:** Since ZenML code runs in any environment, you can 

> develop ML models locally with all your favorite tools. Once you are happy with 

Reviewer 2: Non-sequitur: if you can run your code in any environment, shouldn't the second part of this sentence be about how you can run locally AND in the cloud? (I know you get to that in the next sentence, but it makes this first sentence read funny).

Author: True, how would you formulate it then?

Reviewer 2: ```
**Develop Locally:** ZenML allows you to develop ML models in any environment using your favorite tools. When you're satisfied with your results locally, for example, a single command facilitates your transition to a production environment, where all your previously developed code will continue to function effectively.
```

Maybe?

Author: Great suggestion, thanks

your results, you can switch to a production environment via a single command 
and all your code will just work:

```bash
python run.py  # develop your code locally with all your favorite tools
zenml stack set production
python run.py  # run on production infrastructure without any code changes
```

**Pythonic SDK:** ZenML is designed to be as unintrusive as possible. Adding 
a ZenML `@step` or `@pipeline` decorator to your Python functions is enough to 
turn your existing code into ZenML pipelines:

```python
from zenml import pipeline, step

@step
def step_1() -> str:
  return "world"

@step
def step_2(input_one: str, input_two: str) -> None:
  combined_str = input_one + ' ' + input_two
  print(combined_str)

@pipeline
def my_pipeline():
  output_step_one = step_1()
  step_2(input_one="hello", input_two=output_step_one)

my_pipeline()
```

**Reusability:** Defining your code as ZenML steps brings standardization 

> into your organization, allowing you to easily reuse code developed by your 

Reviewer 2: 'brings standardization into your organization' feels a bit vague...

Author: Rewrote that paragraph entirely

colleagues:

```python
from my_organization.steps import data_loader_step, model_validation_step

@step
def my_model_trainer_step(data):
    ...

@pipeline
def my_pipeline():
  data = data_loader_step()
  model = my_model_trainer_step(data)
  eval_metric = model_validation_step(model)
```
{% endtab %}

{% tab title="ZenML for ML Engineers" %}

> As an ML Engineer, ZenML empowers you to take ownership of the entire ML 

Reviewer 2: See above. 'As an ML Engineer'.

Maybe it's fine. 🤷‍♂️

lifecycle end-to-end. Adopting ZenML means fewer handover points and more 
visibility on what is happening in your organization.

**ML Lifecycle Management:** ZenML's abstractions enable you to manage
sophisticated ML setups with ease. After you define your ML workflows as 
[Pipelines](./getting-started/core-concepts.md#1-development) and your 
development, staging, and production infrastructures as 
[Stacks](./getting-started/core-concepts.md#2-execution), you can move 
entire ML workflows to different environments in seconds.

```bash
zenml stack set staging
python run.py  # test your workflows on staging infrastructure

> zenml stack set prod

Reviewer: maybe change to `production` to be consistent with the mlops tab

python run.py  # run your workflows in production
```

**Reproducibility:** ZenML enables you to painlessly reproduce previous results

> by automatically tracking and versioning stacks, pipelines, and output 

Reviewer 2: it's implied by 'pipelines', but do we need to mention code tracking / versioning here?

Author: Doesn't hurt I guess

artifacts. In the ZenML dashboard, you can get an overview of everything that 
has happened. Try it out at [https://demo.zenml.io/](https://demo.zenml.io/)!

<figure>

> <img src=".gitbook/assets/intro_dashboard.png" alt="ZenML Dashboard Overview" width="50%">

Reviewer 2: can you make this an `<a>` link to the demo dashboard?

Author: You mean at the bottom of the page in the next steps section? Or just here instead of the current link (if so, what's the advantage of using HTML links directly?)

Reviewer 2: No I just mean that clicking on the image should take you to the demoserver... You can wrap the image itself in a link.

Reviewer 2: Add the link to the image, not to the figure?

</figure>

**Automated Deployments:** With ZenML, you no longer need to upload custom 
Docker images to the cloud whenever you want to deploy a new model to 
production. Simply define your ML workflow as a ZenML pipeline, let ZenML 

> handle all the containerization, and have your model automatically deployed to 

Reviewer 2: ```suggestion
handle the containerization, and have your model automatically deployed to 
```

a highly scalable Kubernetes deployment service like 
[Seldon](./user-guide/component-galery/model-deployers/seldon.md).

```python
from zenml.integrations.seldon.steps import seldon_model_deployer_step

from my_organization.steps import data_loader_step, model_trainer_step

@pipeline
def my_pipeline():
  data = data_loader_step()
  model = model_trainer_step(data)
  seldon_model_deployer_step(model)
```

{% endtab %}

{% tab title="ZenML for MLOps / Platform Engineers" %}

> As an MLOps infrastructure expert, ZenML enables you to define, deploy, and 

Reviewer 2: See above.

manage sophisticated production environments that are easy to share with your 
colleagues.

**Standardization:** With ZenML, you can standardize MLOps infrastructure and 
tooling across your organization. Simply set up a ZenML server, register your 
production environment as a ZenML stack, and invite your colleagues to run 
their ML workflows on it:

```bash

> zenml deploy  # Deploy ZenML

Reviewer: We can also add deploying mlflow here as an example. I would even suggest moving the deployment part into a seperate tab altogether so:

```
zenml deploy  # Deploy ZenML on any cloud
zenml experiment-tracker deploy my_tracker --flavor=mlflow --cloud=gcp  # Deploy any MLOps component in seconds
```

```
zenml secret create mlflow --password=my_password  # Create secrets on supported backends
zenml experiment-tracker register my_mlflow_tracker --flavor=mlflow --password {{ mlflow.my_password }}  # register stacks with ease
zenml stack register production...  # Register your production environment
zenml stack share production  # Make it available to your colleagues
```

Author: I added a `zenml ... deploy` example here

Author: Secrets syntax feels a bit too detailed for an intro page IMO

Author: I replaced the extensibility section with a separate deployment section now

zenml stack register production ...  # Register your production environment
zenml stack share production  # Make it available to your colleagues
```

**No Vendor Lock-In:** Since infrastructure is decoupled from code, ZenML gives
you the freedom to switch to a different tooling stack whenever it suits you.
By avoiding vendor lock-in, you have the flexibility to transition between 

> various cloud providers or services, ensuring that you receive the best 

Reviewer 2: ```suggestion
cloud providers or services, ensuring that you receive the best 
```

> performance and pricing available in the market at any time.

Reviewer 2: 🔥🚀
Good line!

Author: hahaha that's the one line I didn't write myself 😆

```bash
zenml stack set gcp
python run.py  # Run your ML workflows in GCP
zenml stack set aws
python run.py  # Now your ML workflow runs in AWS
```

**Extensibility:** ZenML is open-source and highly extensible, so you benefit 

> from industry-standard best-practices built into the framework while also being 

Reviewer 2: ```suggestion
from industry-standard best practices built into the framework while  
```

> able to customize it to fit your specific needs.

Reviewer 2: ```suggestion
retaining the option to customize it to fit your specific needs.
```

{% endtab %}

{% endtabs %}

## Directions

Ready to harness the power of ZenML for your machine learning projects? Here is a collection of pages you can take a look at next:

> <table data-view="cards">

Author: I added some in the Platform tab, doesn't look too bad actually

> <table data-view="cards">

Author: What do you think, should we keep it? Which links would you add to which tab exactly?

> <table data-view="cards">

Reviewer: Lets discuss in person - generally I think now its confusing as there are too many cards :D

# ⭐ Introduction

ZenML is an extensible, open-source MLOps framework for creating portable, 
production-ready machine learning pipelines.
By decoupling infrastructure from code, ZenML enables developers across your 
organization to collaborate more effectively as they develop to production.

![ZenML Overview](.gitbook/assets/intro_zenml_overview.png)

{% tabs %}

{% tab title="ZenML for Data Scientists" %}

ZenML gives data scientists the freedom to fully focus on modeling and

> experimentation while writing code that is immediately production-ready.

Reviewer 2: ```suggestion
experimentation while writing code that is ready for production from the get-go.
```

**Develop Locally:** Since ZenML code runs in any environment, you can 
develop ML models locally with all your favorite tools. Once you are happy with 
your results, you can switch to a production environment via a single command 
and all your code will just work:

```bash
python run.py  # develop your code locally with all your favorite tools
zenml stack set production
python run.py  # run on production infrastructure without any code changes
```

**Pythonic SDK:** ZenML is designed to be as unintrusive as possible. Adding 
a ZenML `@step` or `@pipeline` decorator to your Python functions is enough to 
turn your existing code into ZenML pipelines:

```python
from zenml import pipeline, step

@step
def step_1() -> str:
  return "world"

@step
def step_2(input_one: str, input_two: str) -> None:
  combined_str = input_one + ' ' + input_two
  print(combined_str)

@pipeline
def my_pipeline():
  output_step_one = step_1()
  step_2(input_one="hello", input_two=output_step_one)

my_pipeline()
```

> **Automatic Metadata Tracking** ZenML automatically tracks metadata of all your
> experiments and saves all your datasets and models to disk and versions them.
> Using the ZenML dashboard, you can then see detailed visualizations of all your 
> experiments.

Reviewer 2: ```suggestion
**Automatic Metadata Tracking** ZenML automatically tracks metadata for all your pipeline runs, saving all your datasets and models to disk and versioning them.
Using the ZenML dashboard, you can then see detailed visualizations of all your 
experiments.
```

Maybe? Felt like too many 'experiments'

ZenML enables MLOps infrastructure experts to define, deploy, and manage
sophisticated production environments that are easy to share with colleagues.

**Built-in Deployment:** ZenML can be deployed on any cloud provider and 
provides many Terraform-based utility functions to deploy other MLOps tools or 
even entire MLOps stacks:

```bash
zenml deploy --provider aws  # Deploy ZenML to any cloud
zenml orchestrator deploy kfp --flavor kubeflow --cloud gcp  # Deploy MLOps tools and infrastructure to any cloud
zenml stack recipe deploy gcp-vertexai  # Deploy entire MLOps stacks at once
```

**Standardization:** With ZenML, you can standardize MLOps infrastructure and 
tooling across your organization. Simply register your staging and production 
environments as ZenML stacks and invite your colleagues to run ML workflows on 
them.

```bash
zenml orchestrator register kfp_orchestrator -f kubeflow  # Register MLOps tools and infrastructure
zenml stack register production --orchestrator kubeflow ...  # Register your production environment
zenml stack share production  # Make it available to your colleagues
```

Registering your environments as ZenML stacks also enables you to browse and

> explore all your infrastructure setups in a convenient user interface. 

Reviewer 2: 'infrastructure setups'? if you're saying 'stacks' earlier in the sentence, why do you have to use this new phrasing?

If you answered yes to any of the questions above, ZenML is here to help you:

**ZenML** is an extensible, open-source MLOps framework for creating portable, 
production-ready MLOps pipelines. 

> By decoupling infrastructure from code, ZenML gives all ML developers in your 

Reviewer 3: Just based on the look of the page, I would also merge this paragraph with the previous paragraph.

Author: Makes sense. Let's see how it looks like.

organization more freedom and independence in how they approach their work:

![ZenML Overview](.gitbook/assets/intro_zenml_overview.png)

{% tabs %}

> {% tab title="ZenML for Data Scientists" %}

Reviewer: I would make the first tab the MLOps Platform Engineer

Author: I changed it but personally I liked the DS tab as default tab better since it better showcases what ZenML does at a glance.

Author: Changed it back for that reason

Author: > Well our target audience is the MLOPs platform engineer so it should showcase it if it doesnt.

True but ZenML might be used by people from a variety of backgrounds, maybe you're a CTO, or product manager, or a student learning about MLOps, then the platform tab would not be too informative IMO. If you're literally a platform engineer, wouldn't you click on the respective tab right away?

> How about adding a pipeline with one or two basic steps in the MLOps platform engineer tab and showcasing the stack switching as well?

Stack switching is already there under the vendor lock problem, but I don't think it makes sense to add pipelining features here since the platform engineer is not going to write pipelines, so this might not be relevant to them.

> As a Data Scientists, ZenML gives you the freedom to focus on modeling and

Reviewer: ```suggestion
As a Data Scientist, ZenML gives you the freedom to focus on modeling and
```

> As a Data Scientists, ZenML gives you the freedom to focus on modeling and

Reviewer 3: If I am not mistaken, when you start a sentence with "As a .....", it refers to the subject of the sentence. So, in this context, ZenML becomes the Data Scientist. I would perhaps change it to "As data scientists, you get the freedom .....".

Author: @strickvl /AlexGPT is this true?

Author: Apparently the formulation is not optimal yeah 😄

> experimentation without having to worry about production issues at all.

Reviewer 2: Is it 'you don't have to worry about production', or is it 'we empower you that you can get to do things that otherwise you wouldn't have been able to do'?

Reviewer: I also agree that we should rather focus that we connect you to production

Author: Good point, current formulation is a bit too negative perhaps 👍

> **Tooling Freedom:** Since ZenML code runs in any environment, you can 

Reviewer: Maybe also say here (either add it or instead of ) that we help you create production-ready code *locally*  first without infrastructure worries and give you a way to structure your code properly. These are the values ive heard more often than tooling freedom

Author: Running locally was already mentioned before, but more as a side note. I changed the header to **Develop Locally:** now to emphasize this a bit more.

develop ML models locally with all your favourite tools. Once you are happy with 
your results, you can switch to a production environment via a single command 
and all your code will just work:

```bash
python run.py  # develop your code locally with all your favourite tools
zenml stack set production
python run.py  # run your code in production without any code changes
```

> **Pythonic SDK:** ZenML is designed to be as unintrusive as possible. Adding 

Reviewer 3: Based on the look of the page, how about putting the 3 points in each tab into lists? Something like:

> As a ....., .........
> 
> - Tooling Freedom: ....
> - Reuseability: ...

Author: Yeah I had that before but it looks weird because the items will be indented but the code snippets are not

Reviewer 3: As far as I know, if you like the lists, we can indent the code blocks afterward in GitBook.

a ZenML `@step` or `@pipeline` decorator to your Python functions is enough to 
turn your existing code into ZenML pipelines:

```python
@step
def step_1() -> str:
  return "world"

@step
def step_2(input_one: str, input_two: str) -> None:
  combined_str = input_one + ' ' + input_two
  print(combined_str)

@pipeline
def my_pipeline():
  output_step_one = step_1()
  step_2(input_one="hello", input_two=output_step_one)

my_pipeline()
```

**Reusability:** Defining your code as ZenML steps brings standardization 
into your organization, allowing you to easily reuse code developed by your 
colleagues:

```python
from my_organization.steps import data_loader_step, model_validation_step

@step
def my_model_trainer_step(data):
    ...

@pipeline
def my_pipeline():
  data = data_loader_step()
  model = my_model_trainer_step(data)
  eval_metric = model_validation_step(model)
```
{% endtab %}

{% tab title="ZenML for ML Engineers" %}

As an ML Engineer, ZenML enables you take ownership of the entire ML workflow

> end-to-end. Adopting ZenML means less handover points and more visibility on

Reviewer 2: ```suggestion
end-to-end. Adopting ZenML means fewer handover points and more visibility on
```

what is happening in your organization.

> **ML Lifecycle Management:** ZenML's abstractions make it possible to manage the

Reviewer 3: Not sure if this is possible but can we add an image/code block/CLI commands somewhere in the ML Engineer tab. In comparison to the other tabs, this tab looks a bit too text-heavy.

entire ML lifecycle of your company end-to-end. Complex infrastructure setups 
are bundled into "stacks" and sophisticated ML workflows into "pipelines",
enabling you to move ML workflows between different environments in seconds.

**Reproducibility:** ZenML enables you to painlessly reproduce previous results
by automatically tracking and versioning stacks, pipelines, and output 
artifacts. In the ZenML dashboard you can get an overview of everything that has happened. Try it out on https://demo.zenml.io/!

**Automated Deployments:** With ZenML you no longer need to upload custom 
Docker images to the cloud every time you want to deploy a new model to 
production. Simply define your ML workflow as a ZenML pipeline, let ZenML 
handle all the containerization, and have your model automatically deployed to 
a highly scalable Kubernetes deployment service like 
[Seldon](./user-guide/component-galery/model-deployers/seldon.md).

{% endtab %}

{% tab title="ZenML for MLOps / Platform Engineers" %}

> As an MLOps infrastructre expert, ZenML enables you to define, deploy, and 

Reviewer: Maybe just say As an MLOps Platform Engineer?

Reviewer: Infrastructure is spelled wrong

manage sophisticated production environments that are easy to share with your 
colleagues.

**Standardization:** With ZenML, you can standardize MLOps infrastructure and 
tooling across your organization. Simply set up a ZenML server, register your 
production environment as a ZenML stack, and invite your colleagues to run 
their ML workflows on it:

```bash
zenml deploy  # Deploy ZenML
zenml stack register production ...  # Register your production environment
zenml stack share production  # Make it available to your colleagues
```

> **No Vendor Lock:** Since infrastructure is decoupled from code, you will never 

Reviewer: No Vendor Lock-in*?

> be vendor-locked with ZenML. ZenML gives you the freedom to switch to a 

Reviewer: Maybe also mention running workloads on cloud more efficiently reduces cost?

Author: Okay I GPT-ed something up, WDYT?

> different tooling stack whenever it suites you:

Reviewer 2: ```suggestion
different tooling stack whenever it suits you:
```

```bash
zenml stack set gcp
python run.py  # Run your ML workflows in GCP
zenml stack set aws
python run.py  # Now your ML workflow runs in AWS
```

**Extensibility:** ZenML is open-source and highly extensible, so you benefit 
from industry-standard best-practices built into the framework while also being 
able to customize it to fit your specific needs.

{% endtab %}

{% endtabs %}

## Directions

Ready to harness the power of ZenML for your machine learning projects? Here is a collection of pages you can take a look at next:

> <table data-view="cards"><thead><tr><th></th><th></th><th></th><th data-hidden data-card-target data-type="content-ref"></th></tr></thead><tbody><tr><td><span data-gb-custom-inline data-tag="emoji" data-code="1f4bf">💿</span> <strong>Installation</strong></td><td>Install ZenML to get started.</td><td></td><td><a href="getting-started/installation.md">installation.md</a></td></tr><tr><td><span data-gb-custom-inline data-tag="emoji" data-code="1f9f1">🧱</span> <strong>Core Concepts</strong></td><td>Understand the core concepts behind ZenML.</td><td></td><td><a href="getting-started/core-concepts.md">core-concepts.md</a></td></tr><tr><td><span data-gb-custom-inline data-tag="emoji" data-code="1f423">🐣</span> <strong>Starter Guide</strong></td><td>Get started quickly with a simple setup.</td><td></td><td></td></tr><tr><td><span data-gb-custom-inline data-tag="emoji" data-code="1f414">🐔</span> <strong>Advanced Guide</strong></td><td>Hone your MLOps skills with an advanced guide.</td><td></td><td></td></tr><tr><td><span data-gb-custom-inline data-tag="emoji" data-code="1f4cb">📋</span> <strong>Component Guide</strong></td><td>Browse through the already-implemented integrations for ZenML.</td><td></td><td><a href="user-guide/component-galery/">component-galery</a></td></tr><tr><td><span data-gb-custom-inline data-tag="emoji" data-code="1f3d7">🏗</span> <strong>Set up your MLOps Platform</strong></td><td>Establish your production-ready infrastructure.</td><td></td><td><a href="platform-guide/set-up-your-mlops-platform/">set-up-your-mlops-platform</a></td></tr><tr><td><span data-gb-custom-inline data-tag="emoji" data-code="1f9d1-1f692">🧑🚒</span> <strong>Manage the Deployed Services</strong></td><td>Manage your production-ready infratstructure.</td><td></td><td><a href="platform-guide/manage-the-deployed-services/">manage-the-deployed-services</a></td></tr><tr><td><span data-gb-custom-inline data-tag="emoji" data-code="1f9d1-1f3eb">🧑🏫</span> <strong>Examples</strong></td><td>Take a peek at how ZenML works in concrete examples.</td><td></td><td><a href="learning/examples/">examples</a></td></tr><tr><td><span data-gb-custom-inline data-tag="emoji" data-code="1f64b">🙋</span> <strong>FAQ</strong></td><td>Find the answers to the most frequently asked questions.</td><td></td><td><a href="learning/faq.md">faq.md</a></td></tr></tbody></table>

Reviewer: Personally, I feel like these are too many directions. If we have 9 i would some put them in the tabs, so 3 for each persona (ideally they would be linked to the values we provide, or a starting guide for them).

Alternatively (or additionally), give only 3 directions: User guide, Platform Guide, and Core Concepts for all personas

Reviewer 3: With the current page, I agree with @htahir1 on this one. 9 directions might be a bit too much. However, considering that this is the introduction page, I would propose a different set of three pages: Quickstart, Core Concepts, Installation(or Starter Guide) in any order.

Author: Actually I like the current layout. I think having those tiles looks quite clean and is not overwhelming either. And it's a great first docs page where you can find quick links to what you need at a glance.

Adding persona-specific next steps could be interesting but I'm not 100% sure if you can split it cleanly and it could also hide relevant links below tabs that the user might not click.

Author: @bcdurak oh yeah the quickstart is missing from the tiles right now, good point. I just added it 👍
